File: yangxj96-starter/yangxj96-starter-remote/src/main/java/io/github/yangxj96/starter/remote/autoconfigure/RemoteAutoConfiguration.java
package io.github.yangxj96.starter.remote.autoconfigure;

import feign.*;
import feign.okhttp.OkHttpClient;
import io.github.yangxj96.starter.remote.props.RemoteProperties;
import jakarta.annotation.PreDestroy;
import lombok.extern.slf4j.Slf4j;
import okhttp3.ConnectionPool;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.cloud.openfeign.EnableFeignClients;
import org.springframework.cloud.openfeign.FeignAutoConfiguration;
import org.springframework.cloud.openfeign.loadbalancer.FeignLoadBalancerAutoConfiguration;
import org.springframework.cloud.openfeign.support.FeignHttpClientProperties;
import org.springframework.cloud.openfeign.support.PageJacksonModule;
import org.springframework.cloud.openfeign.support.SortJacksonModule;
import org.springframework.cloud.openfeign.support.SpringMvcContract;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;

import java.util.concurrent.TimeUnit;

/**
 * 远程请求的openfeign配置
 */
@Slf4j
@Import(value = {
        FeignLoadBalancerAutoConfiguration.class,
        FeignAutoConfiguration.class
})
@AutoConfiguration(before = FeignLoadBalancerAutoConfiguration.class)
@EnableFeignClients("io.github.yangxj96.starter.remote.clients")
@EnableConfigurationProperties(RemoteProperties.class)
public class RemoteAutoConfiguration {

    private static final String LOG_PREFIX = "[自动配置-远程调用]:";

    /**
     * 项目配置文件
     */
    private final RemoteProperties properties;

    public RemoteAutoConfiguration(@Autowired RemoteProperties properties) {
        this.properties = properties;
    }

    /**
     * feign 请求超时时间配置
     * feign 日志级别
     *
     * @return feign 日志
     */
    @Bean
    public Logger.Level level() {
        return properties.getLevel();
    }

    /**
     * 契约配置
     *
     * @return 契约
     */
    @Bean
    public Contract contract() {
        // 默认契约
        // return new Contract.Default()
        // spring 包装好的契约
        return new SpringMvcContract();
    }


    /**
     * feign超时时间
     *
     * @return Options
     */
    @Bean
    public Request.Options options() {
        return new Request.Options(
                // @formatter:off
                properties.getConnectTimeOut(), TimeUnit.MILLISECONDS,
                properties.getReadTimeOut()   , TimeUnit.MILLISECONDS,
                true);
                // @formatter:on
    }


    /**
     * feign 重试机制
     *
     * @return 重试机制
     */
    @Bean
    public feign.Retryer retryer() {
        return new Retryer.Default(100, TimeUnit.SECONDS.toMillis(1), 2);
    }

    @Bean
    public RequestInterceptor requestInterceptor() {
        return template -> template.header("feign", "true");
    }

    /////////////////////////////// okhttp3
    private okhttp3.OkHttpClient okHttpClient;

    @Bean
    @ConditionalOnMissingBean
    public okhttp3.OkHttpClient.Builder okHttpClientBuilder() {
        return new okhttp3.OkHttpClient.Builder();
    }

    @Bean
    public ConnectionPool httpClientConnectionPool(FeignHttpClientProperties httpClientProperties) {
        int maxTotalConnections = httpClientProperties.getMaxConnections();
        long timeToLive = httpClientProperties.getTimeToLive();
        TimeUnit ttlUnit = httpClientProperties.getTimeToLiveUnit();
        return new ConnectionPool(maxTotalConnections, timeToLive, ttlUnit);
    }

    @Bean
    public okhttp3.OkHttpClient okHttpClient(okhttp3.OkHttpClient.Builder builder, ConnectionPool connectionPool) {
        this.okHttpClient = builder
                .connectTimeout(properties.getConnectTimeOut(), TimeUnit.MILLISECONDS)
                .followRedirects(true)
                .readTimeout(properties.getReadTimeOut(), TimeUnit.MILLISECONDS)
                .connectionPool(connectionPool)
                .build();
        return this.okHttpClient;
    }

    @PreDestroy
    public void destroy() {
        if (this.okHttpClient != null) {
            this.okHttpClient.dispatcher().executorService().shutdown();
            this.okHttpClient.connectionPool().evictAll();
        }
    }

    @Bean
    @ConditionalOnMissingBean(Client.class)
    public Client feignClient(okhttp3.OkHttpClient client) {
        return new OkHttpClient(client);
    }


    /////////////////////////////// jackson

    @Bean
    public PageJacksonModule pageJacksonModule() {
        return new PageJacksonModule();
    }

    @Bean
    public SortJacksonModule sortModule() {
        return new SortJacksonModule();
    }

    ///////////////////////////////  circuitbreaker
//
//    @Bean
//    public Targeter defaultFeignTargeter() {
//        return new Targeter() {
//            @Override
//            public <T> T target(FeignClientFactoryBean factory, Feign.Builder feign, FeignClientFactory context, Target.HardCodedTarget<T> target) {
//                return feign.target(target);
//            }
//        };
//    }
//
//    @Bean
//    public CircuitBreakerNameResolver alphanumericCircuitBreakerNameResolver() {
//        return (feignClientName, target, method) -> Feign.configKey(target.type(), method).replaceAll("[^a-zA-Z0-9]", "");
//    }
//
//    @SuppressWarnings("rawtypes")
//    @Bean
//    public Targeter circuitBreakerFeignTargeter(CircuitBreakerFactory circuitBreakerFactory,
//                                                @Value("${spring.cloud.openfeign.circuitbreaker.group.enabled:false}") boolean circuitBreakerGroupEnabled,
//                                                CircuitBreakerNameResolver circuitBreakerNameResolver) {
//        return new FeignCircuitBreakerTargeter(circuitBreakerFactory, circuitBreakerGroupEnabled, circuitBreakerNameResolver);
//    }


}

File: yangxj96-starter/yangxj96-starter-remote/src/main/java/io/github/yangxj96/starter/remote/autoconfigure/RemoteAutoConfiguration.kt
package io.github.yangxj96.starter.remote.autoconfigure

import feign.*
import io.github.yangxj96.starter.remote.configure.OkHttpLogInterceptor
import io.github.yangxj96.starter.remote.props.RemoteProperties
import jakarta.annotation.PreDestroy
import jakarta.annotation.Resource
import okhttp3.ConnectionPool
import okhttp3.OkHttpClient
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.AutoConfiguration
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.boot.web.servlet.FilterRegistrationBean
import org.springframework.cloud.client.loadbalancer.LoadBalanced
import org.springframework.cloud.client.loadbalancer.LoadBalancerClient
import org.springframework.cloud.loadbalancer.blocking.client.BlockingLoadBalancerClient
import org.springframework.cloud.loadbalancer.support.LoadBalancerClientFactory
import org.springframework.cloud.openfeign.EnableFeignClients
import org.springframework.cloud.openfeign.loadbalancer.FeignBlockingLoadBalancerClient
import org.springframework.cloud.openfeign.loadbalancer.FeignLoadBalancerAutoConfiguration
import org.springframework.cloud.openfeign.loadbalancer.LoadBalancerFeignRequestTransformer
import org.springframework.cloud.openfeign.support.FeignHttpClientProperties
import org.springframework.cloud.openfeign.support.PageJacksonModule
import org.springframework.cloud.openfeign.support.SortJacksonModule
import org.springframework.cloud.openfeign.support.SpringMvcContract
import org.springframework.context.annotation.Bean
import org.springframework.core.annotation.Order
import java.security.KeyManagementException
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager


@Order()
@EnableFeignClients("io.github.yangxj96.starter.remote.clients")
@AutoConfiguration(before = [FeignLoadBalancerAutoConfiguration::class])
@EnableConfigurationProperties(RemoteProperties::class)
class RemoteAutoConfiguration(@Resource private val properties: RemoteProperties) {

    companion object {
        private const val PREFIX = "[自动配置-远程调用]:"

        private val log = LoggerFactory.getLogger(this::class.java)
    }

    private lateinit var client: OkHttpClient

    /**
     * feign 请求超时时间配置
     * feign 日志级别
     *
     * @return feign 日志
     */
    @Bean
    fun level(): Logger.Level {
        log.debug("$PREFIX 配置feign日志级别")
        return properties.level
    }

    /**
     * 契约配置
     *
     * @return 契约
     */
    @Bean
    fun contract(): Contract {
        log.debug("$PREFIX 配置feign契约")
        // spring 包装好的契约
        return SpringMvcContract()
    }

    /**
     * feign超时时间
     *
     * @return Options
     */
    @Bean
    fun options(): Request.Options {
        log.debug("$PREFIX 配置feign超时时间")
        return Request.Options(
            properties.connectTimeOut, TimeUnit.MILLISECONDS,
            properties.readTimeOut, TimeUnit.MILLISECONDS,
            true
        )
    }

    /**
     * feign 重试机制
     *
     * @return 重试机制
     */
    @Bean
    fun retryer(): Retryer {
        log.debug("$PREFIX 配置feign重试机制")
        return properties.retryer
    }

    /**
     * openfeign请求拦截器
     */
    @Bean
    fun requestInterceptor(): RequestInterceptor {
        log.debug("$PREFIX 配置feign请求拦截器")
        return RequestInterceptor { template: RequestTemplate -> template.header("feign", "true") }
    }

    /////////////////////////////// okhttp3

    /**
     * http客户端连接池
     */
    @Bean
    fun httpClientConnectionPool(httpClientProperties: FeignHttpClientProperties): ConnectionPool {
        log.debug("$PREFIX 配置feign okhttp连接池")
        // 最大连接数,默认活动时间,活动时间单位
        return ConnectionPool(200, 900L, TimeUnit.SECONDS)
    }

    /**
     * okhttp客户端
     * @param connectionPool 连接池
     */
    @Bean
    @LoadBalanced
    fun okHttpClient(connectionPool: ConnectionPool): OkHttpClient {
        log.debug("$PREFIX 配置feign okhttp客户端")
        val build = OkHttpClient.Builder()
            .connectTimeout(properties.connectTimeOut, TimeUnit.MILLISECONDS)
            .readTimeout(properties.readTimeOut, TimeUnit.MILLISECONDS)
            .writeTimeout(properties.writeTimeout, TimeUnit.MILLISECONDS)
            .addInterceptor(OkHttpLogInterceptor())
            .connectionPool(connectionPool)

        try {
            val disabledTrustManager: X509TrustManager = object : X509TrustManager {
                override fun checkClientTrusted(p0: Array<out X509Certificate>?, p1: String?) {

                }

                override fun checkServerTrusted(p0: Array<out X509Certificate>?, p1: String?) {

                }

                override fun getAcceptedIssuers(): Array<X509Certificate?> {
                    return arrayOfNulls(0)
                }

            }
            val trustManagers = arrayOfNulls<TrustManager>(1)
            trustManagers[0] = disabledTrustManager
            val sslContext = SSLContext.getInstance("SSL")
            sslContext.init(null, trustManagers, SecureRandom())
            val disabledSSLSocketFactory = sslContext.socketFactory
            build.sslSocketFactory(disabledSSLSocketFactory, disabledTrustManager)
            build.hostnameVerifier { _, _ -> true }

        } catch (e: NoSuchAlgorithmException) {
            log.warn("Error setting SSLSocketFactory in OKHttpClient", e)
        } catch (e: KeyManagementException) {
            log.warn("Error setting SSLSocketFactory in OKHttpClient", e)
        }

        this.client = build.build()
        return this.client
    }

    @PreDestroy
    fun destroy() {
        log.debug("$PREFIX 准备销毁okhttp")
        if (this::client::isInitialized.get()) {
            this.client.dispatcher.executorService.shutdown()
            this.client.connectionPool.evictAll()
        }
    }

    @Bean
    fun loadBalancedClient(loadBalancerClientFactory: LoadBalancerClientFactory?): LoadBalancerClient {
        return BlockingLoadBalancerClient(loadBalancerClientFactory)
    }

    /**
     * 构建feign客户端
     */
    @Bean
    fun feignClient(
        client: OkHttpClient,
        loadBalancedClient: LoadBalancerClient,
        loadBalancerClientFactory: LoadBalancerClientFactory,
        transformers: List<LoadBalancerFeignRequestTransformer>
    ): FeignBlockingLoadBalancerClient {
        log.debug("$PREFIX 配置feign okhttp的客户端")
        val delegate = feign.okhttp.OkHttpClient(client)
        return FeignBlockingLoadBalancerClient(delegate, loadBalancedClient, loadBalancerClientFactory, transformers)
    }


    /////////////////////////////// jackson
    @Bean
    fun pageJacksonModule(): PageJacksonModule {
        log.debug("$PREFIX 配置feign中jackson适配的PageJacksonModule")
        return PageJacksonModule()
    }

    @Bean
    fun sortModule(): SortJacksonModule {
        log.debug("$PREFIX 配置feign中jackson适配的SortJacksonModule")
        return SortJacksonModule()
    }

}
